### PR TITLE
Feat (Vulnerability): Improve add/update/delete vulnerability APIs, implement add/update/delete vulnerability by GUI

### PR DESCRIPTION
**Pull request by hoangnt2**
Feat (Vulnerability): Improve add/update/delete vulnerability APIs implement add/update/delete vulnerability by GUI

Issue: 
#1170 

### 1. Implement create, update, delete vulnerability by GUI
Refer to #1170

### 2. Improvement vulnerability REST API
Change format cveReferences field in request body of add and update vulnerability REST API to set<string> ("year-number")

* Create a vulnerability

```
curl 'https://sw360.org/api/vulnerabilities' -i -X POST -H 'Content-Type: application/hal+json'
 -H 'Authorization: Bearer eyJhbGciOi*******' -H 'Accept: application/hal+json' 
 -d '{
 "externalId" : "987",
 "title" : "Title of vulnerability 12345",
 "description" : "Description of vulnerability",
 "publishDate" : "01-01-2018",
 "lastExternalUpdate" : "last-update-id-123",
 "priority" : "high",
 "priorityText" : "Priority Text",
 "action" : "new",
 "impact" : {
   "Key 1" : "Value 1",
   "Key 2" : "Value 2"
 },
 "legalNotice" : "Legal Notice",
 "assignedExtComponentIds" : [ "comp-id-2", "comp-id-1" ],
 "cveReferences" : [ 
     "2021-1111",
     "2022-2222"
  ],
 "vendorAdvisories" : [ {
   "vendor" : "vendor1",
   "name" : "name1",
   "url" : "http://url1"
 }, {
   "vendor" : "vendor2",
   "name" : "name2",
   "url" : "http://url2"
 } ],
 "extendedDescription" : "Extended Description",
 "references" : [ "reference-1", "reference-2" ],
 "cvss" : "3.0",
 "cvssTime" : "2019-01-01",
 "vulnerableConfiguration" : {
   "Key 1" : "Value 1",
   "Key 2" : "Value 2"
 },
 "access" : {
   "Key 1" : "Value 1",
   "Key 2" : "Value 2"
 },
 "cwe" : "common weakness enumeration",
 "cveFurtherMetaDataPerSource" : {
   "Cve meta data" : {
     "Key 1" : "Value 1",
     "Key 2" : "Value 2"
   }
 }
}'

```

* Update vulnerability 
```
curl 'https://sw360.org/api/vulnerabilities/98765' -i -X PATCH -H 'Content-Type: application/hal+json' -H 'Authorization: Bearer eyJhbGciOiJSUzI1N*********' -H 'Accept: application/hal+json' -d '{
  "externalId" : "98765",
  "title" : "Updated Title",
  "description" : "Description of vulnerability",
  "publishDate" : "01-01-2018",
  "lastExternalUpdate" : "last-update-id-123",
  "priority" : "high",
  "priorityText" : "Priority Text",
  "action" : "new",
  "impact" : {
    "Key 1" : "Value 1",
    "Key 2" : "Value 2"
  },
  "legalNotice" : "Legal Notice",
  "assignedExtComponentIds" : [ "comp-id-2", "comp-id-1" ],
  "cveReferences" : [ 
     "2021-1111",
     "2022-2222"
  ],
  "vendorAdvisories" : [ {
    "vendor" : "vendor1",
    "name" : "name1",
    "url" : "http://url1"
  }, {
    "vendor" : "vendor2",
    "name" : "name2",
    "url" : "http://url2"
  } ],
  "extendedDescription" : "Extended Description",
  "references" : [ "reference-1", "reference-2" ],
  "cvss" : "3.0",
  "cvssTime" : "2019-01-01",
  "vulnerableConfiguration" : {
    "Key 1" : "Value 1",
    "Key 2" : "Value 2"
  },
  "access" : {
    "Key 1" : "Value 1",
    "Key 2" : "Value 2"
  },
  "cwe" : "common weakness enumeration",
  "cveFurtherMetaDataPerSource" : {
    "Cve meta data" : {
      "Key 1" : "Value 1",
      "Key 2" : "Value 2"
    }
  }
}'
```
* Other method refer to: http://localhost:8080/resource/docs/api-guide.html#resources-vulnerabilities

### Checklist
Must:
- [ ] All related issues are referenced in commit messages and in PR

### CODE REVIEW

**Comment by nikkuma7** [comment]
Test successful.

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/vulnerabilities/VulnerabilitiesPortlet.java [hunk 1]
```diff
@@ -195,19 +230,352 @@ private void prepareDetailView(RenderRequest request, RenderResponse response) t
...
+    public void removeParamUrl(ActionRequest request, ActionResponse response) {
+        try {
+            String portletId = (String) request.getAttribute(WebKeys.PORTLET_ID);
+            ThemeDisplay tD = (ThemeDisplay) request.getAttribute(WebKeys.THEME_DISPLAY);
+            long plid = tD.getPlid();
+            LiferayPortletURL redirectUrl = PortletURLFactoryUtil.create(request, portletId, plid, PortletRequest.RENDER_PART);
+            request.setAttribute(WebKeys.REDIRECT, redirectUrl.toString());
+            response.sendRedirect(redirectUrl.toString());
+        } catch (IOException e) {
+            log.info("Error remove param url: {}", e.getMessage());
```
Comment (smrutis1) [inline frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/vulnerabilities/VulnerabilitiesPortlet.java:369]: Please make it `log.error`

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/vulnerabilities/VulnerabilitiesPortlet.java [hunk 2]
```diff
@@ -195,19 +230,352 @@ private void prepareDetailView(RenderRequest request, RenderResponse response) t
...
+        }
+    }
+
+    public static void updateVulnerabilityFromRequest(PortletRequest request, Vulnerability vulnerability) throws ParseException {
```
Comment (smrutis1) [inline frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/vulnerabilities/VulnerabilitiesPortlet.java:373]: Please move this to backend

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/vulnerabilities/VulnerabilitiesPortlet.java [hunk 3]
```diff
@@ -195,19 +230,352 @@ private void prepareDetailView(RenderRequest request, RenderResponse response) t
...
+        }
+        if (formatDateTime.parse(cvssDateTime) != null) {
+            vulnerability.setCvssTime(cvssDateTime);
+        }
+        if (formatDateTime.parse(externalUpdateDateTime) != null) {
+            vulnerability.setLastExternalUpdate(externalUpdateDateTime);
+        }
+    }
+
+    public static RequestStatus deleteVulnerability(PortletRequest request, Logger log) {
```
Comment (smrutis1) [inline frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/vulnerabilities/VulnerabilitiesPortlet.java:533]: Please move this to backend

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/JacksonCustomizations.java [hunk 4]
```diff
@@ -770,6 +772,86 @@ static abstract class VulnerabilityMixin extends Vulnerability {
             @JsonProperty("id")
             abstract public String getId();
         }
+        @JsonInclude(JsonInclude.Include.NON_EMPTY)
+        @JsonIgnoreProperties({
+                "id",
+                "revision",
+                "type",
+                //"publishDate",
```
Comment (smrutis1) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/JacksonCustomizations.java:780]: please remove the commented lines

**Review by smrutis1 — CHANGES_REQUESTED** [review]
some changes requested

**Comment by KoukiHama** [comment]
We will update soon, Thank you for reviewing.

**Review by smrutis1 — APPROVED** [review]
Code looks ok

**Comment by hoangnt2** [comment]
@smrutis1  Can you remove **needs code review** label?

**Pull request merged**